Add streamer tutorial (#121)
KULeuven-MICAS/snitch_cluster: docs/tutorial/streamer_design.md

# Streamer

The SNAX streamer is a Chisel-genearted module to help streamline the delivery of data from memory to the accelerator. There is a [detailed streamer documentation](../../hw/chisel/doc/streamer.md), but in this tutorial section we will only cover the high-level aspects and how to configure the streamer.

# Why Do We Need a Streamer?

Accelerators attain peak performance when data continuously streams into them; otherwise, they spend cycles waiting for data availability. Delays often occur due to mismatched data arrangements in shared memory or congestion among accelerators accessing the same banks concurrently.

It's crucial to differentiate between the data layout in memory and the access pattern of an accelerator. The figure below shows two different memory layouts and how an accelerator would get the data. The memory address on the top-right corner is a guide to show the adresses of each data element. Assume each column represents a separate memory bank and each block signifies a data element, with each bank having only one read and one write access port.

> ![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/4428f8d7-2d35-4605-8bec-92929d195643)

Author: MARK TODO!

For example, if the data is arranged in memory layout 1, then input A needs to get data in addresses `[0, 4, 8, 12]` and in the exact order. That means the starting `base_address_a=0` and skip counts with `temporal_stride_a=4`. The `target_address_a` is computed in a simple loop:

```
for(i = 0; i < 4; i++):
  target_address_a = base_address_a + temporal_stride_a*i;
```

Another arrangement in memory layout 2 shows that the data can be arranged in a contiguous fashion. This time, input A needs to access the data in the address sequence `[0,1,2,3]`. Here we set `base_address_a=0` and `temporal_stride_a=1`.

A more complicated example is when a streamer can get multiple data in parallel. This necessitates that we need to also have an address generation that can do spatial parallelism. Particularly it would be convenient to provide a base address and a stride but have all other ports automatically increment per accelerator port. The figure below shows an example accelerator that takes in 3 inputs in parallel and also produces 3 outputs in parallel:

> ![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/2ecaf54a-ae62-42ab-be55-b1cd3c02c377)

Reviewer: Now there are multiple memory layouts on this page named layout 1, maybe consider changing to data layout 3..5?

Author: MARK TODO!

                            sb_1*spatial_stride[1] + sb_0*spatial_stride[0]

          target_address = base_address + temporal_address + spatial_address
```

Where `temporal_bound[*]` pertains to the bounds of the temporal loops and `tb_*` pertains to the temporal indices. The `spatial_bound[*]` peratins to the bounds of the spatial loops and `sb_*` pertains to the spatial indices. Wroup them according to `temporal_address` and `spatial_address` and add them to the `base_address` which results in the `target_address`.

The affine address generation is the working principle of the SNAX streamer. With this, we can flexibily access data in various places of the memory. 

# Streamer Microarchitecture

The figure below shows a more detailed architecture of the streamer. The **(1) streamer** sits between the TCDM interconnect and the accelerator. There is also a **(2) streamer wrapper** to re-wire the Chisel-generated signals. More details of the wrappers are in [Connect The Shell](./connect_shell.md) section.

![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/37900d11-504d-4659-ba9a-aa7efe589975)

## Streamer Interfaces

### **(3) TCDM interface** 

> The **TCDM interface** has request and response channels and also use a decoupled interface. The table below describe the details of the request and response channels:

Reviewer: ```suggestion
The **TCDM interface** has request and response channels and also uses a decoupled interface. The table below describe the details of the request and response channels:
```

| signal name        | description                                |
| :-----------------:| :----------------------------------------: |
| tcdm_req_addr_o    | Target address in memory.                  |
| tcdm_req_data_o    | Data to write in memory.                   |
| tcdm_req_write_o   | Write enable signal. 1 = write, 0 = read.  |
| tcdm_req_strb_o    | Byte strobe.                               |
| tcdm_req_core_id_o | Core ID. Currently unused.                 |

> | tcdm_req_is_core_o | Request from core. core. Currently unused. |

Reviewer: ```suggestion
| tcdm_req_is_core_o | Request from core. Currently unused. |
```

| tcdm_req_valid_o   | Request channel is valid.                  |
| tcdm_req_ready_i   | Request channel is ready.                  |
| tcdm_rsp_data_i    | Response data coming from memory.          |
| tcdm_rsp_valid_i   | Response data is valid.                    |

!!! note

>   Some of the signals are currently unused by the streamer. However, we needed to comply with the TCDM IP of the Snitch platform.

Reviewer: Now I'm wondering why they are not used? :smile:

Author: Good question! Technically some of there weren't even used in the original Snitch repo ^^;

I think the others were meant for transactions between clusters. But locally not. For example:

https://github.com/KULeuven-MICAS/snitch_cluster/blob/26427629e2dec6d8911aa06f3359feca64ba085f/hw/snitch_cluster/src/snitch_cluster.sv#L661-L663

I think it's just an added feature if you want to use it. So for now, we aren't using it but we could later!

Recall that from the `snax-alu.hjson`, the `snax_acc_cfg` contains a `snax_streamer_cfg` which references a streamer template. The `snax_acc_cfg` is shown below for reference:

```hjson
snax_acc_cfg: {
  snax_acc_name: "snax_alu"
  snax_module: "snax_alu_wrapper",
  snax_tcdm_ports: 16,
  snax_num_rw_csr: 3,
  snax_num_ro_csr: 2,
  snax_streamer_cfg: {$ref: "#/snax_alu_streamer_template" }
},
```

If you scroll down to the bottom of the configuration file, you should see the `snax_alu_streamer_template` dictionary set. This template describes the streamer configurations. Let's go through each component.

The `temporal_addrgen_unit_params` component with `loop_dim` generates the number of temporal loop bounds. Since it's just 1, then we only have 1 temporal loop bound.
```hjson
temporal_addrgen_unit_params: {
  loop_dim: 1,
}
```

The `fifo_reader_params` and `fifo_writer_params` configure the FIFOs. Take note that the FIFOs are directly connected to the accelerator interface. It consists of the `fifo_width` which configures the datawidth of the accelerator interface ports. It also has the FIFO depths `fifo_depth` which configures the depth of the FIFOs.

> ```

Reviewer: ```suggestion
```hjson
```

fifo_reader_params: {
  fifo_width: [256, 256],
  fifo_depth: [8, 8],
}

fifo_writer_params: {
  fifo_width: [512],
  fifo_depth: [8],
}
```

The `data_reader_params` and `data_writer_params` effectively configures the TCDM interface ports. It consists of `tcdm_ports_num` which control how many TCDM ports are needed. The `spatial_bounds` pertain to the spatial unrolling factors (your parfor) for each data mover. The `spatial_dim` is the dimension of spatial unrolling factors (your parfor) for each data mover. The `element_width` is the single data element width for each data mover, useful for generating spatial addresses.

> ```

Reviewer: ```suggestion
```hjson
```

data_reader_params:{
  tcdm_ports_num: [4, 4],
  spatial_bounds: [[4], [4]],
  spatial_dim: [1,1],
  element_width: [64,64],
}

data_writer_params:{
  tcdm_ports_num: [8],
  spatial_bounds: [[8]],
  spatial_dim: [1],
  element_width: [64],
}
```

The number of elements in a list pertain to how many data movers there are. For example, there are 2 elements in the `data_reader_params` and therefore it instantiates two read data movers. There is only 1 element in the `data_writer_params` and hence only instantiates 1 write data mover.

Finally there is the `stationarity` configuration which is for stationarity for each data mover. If the stationarity bit is set, the innermost loop for that data mover is set to 1. This does not instantiate anything but rather affects the loop bound for a data mover. Basically, it only fixes the loop bound to 1. This is a special case scenario only. In the SNAX ALU, this is 0. 

## SNAX ALU Streamer CSRs

Based on the configuration file we expect the CSR registers to have:

- 1 register for a single loop bound as specified in the `temporal_addrgen_unit_params -> loop_dim = 1`. If the number is 2 then we would have 2 loop bounds.

> - 3 temporal stride registers, 3 spatial stride registers, and 3 base addres spointers since we have 3 data movers in total (2 read data movers and 1 write data movers).

Reviewer: ```suggestion
- 3 temporal stride registers, 3 spatial stride registers, and 3 base address pointers since we have 3 data movers in total (2 read data movers and 1 write data mover).
```

- 1 start register to start the streamer.
- 1 performance counte register for how many cycles the streamer was active.

Below is a tabulated version of the CSRs of the streamers with the register address, type, and description:

> |  register name      |  register addr  |   type  |                   description                        |

Reviewer: ```suggestion
|  register name      |  register offs  |   type  |                   description                        |
```
Is this not an offset on a base address?

Author: Good point! Yes~

| :-----------------: | :-------------: | :-----: |:---------------------------------------------------- |
| temporal loop bound |       0         |   RW    | Temporal loop bound for address generation           |
| temporal stride 0   |       1         |   RW    | Temporal stride for input A                          |
| temporal stride 1   |       2         |   RW    | Temporal stride for input B                          |
| temporal stride 2   |       3         |   RW    | Temporal stride for output OUT                       |
| spatial stride 0    |       4         |   RW    | Spatial stride for input A                           |
| spatial stride 1    |       5         |   RW    | Spatial stride for input B                           |     
| spatial stride 2    |       6         |   RW    | Spatial stride for output OUT                        |     
| base addr 0         |       7         |   RW    | Base address for input A                             |
| base addr 1         |       8         |   RW    | Base address for input A                             |
| base addr 2         |       9         |   RW    | Base address for output OUT                          |

> | start               |       10        |   RW    | Start register to send configurations                |

Reviewer: What does 
>  Start register to send configurations

mean? Does it start the "streaming?"

Author: Yes!

> | perf. counter       |       11        |   RO    | Performance counter indicating number of cycles ran  | 

Reviewer: > Performance counter indicating number of cycles ran

How is this different from `mcycle`?

Author: `mcycle` only tags the current cycle count from boot up. So technically it just measures the current cycle time from boot up.

performance counters are also more accurate because they directly measures the "busy" or "active" time of an accelerator. If we just do the cycle tagging time from the traces, we also capture cycles caused by setup time like branch conditions or such.

3 - Wait a while since this generates the CSR manager, streamer, and all other wrappers.

4 - When finished, navigate to `/repo/target/snitch_cluster/generated/snax_alu/`

5 - Open the file `snax_alu_streamer_StreamerTop.sv`. This is the Chisel-generated file. It looks like a synthesized netlist. All modules required for the streamer are declared in this file. 

6 - Find the top module `snax_alu_streamer_StreamerTop` within the `snax_alu_streamer_StreamerTop.sv` file. Can you identify the signals discussed in this section?

- Which are the interfaces for the accelerator side?
- Can you identify which are for the write and read ports?
- How many are for write ports and how many are for read ports?
- Which are the interfaces for the TCDM side? How many TCDM ports are there?

7 - In the same `snax_alu_streamer_StreamerTop.sv` file find the `snax_alu_streamer_CsrManager` module. This is the CSR manager of the streamer.

- Can you verify and count how many RW and RO ports there are?
- Why do you think the start register is missing in the listed ports?
- Which CSR number do you think pertains to the spatial stride for input A?

> 7 - Find the generated wrapper `snax_alu_streamer_wrapper.sv`. 

Reviewer: ```suggestion
8 - Find the generated wrapper `snax_alu_streamer_wrapper.sv`. 
```

 - Can you see where the Chisel-generated streamer is instanced?

>  - Can you tell what the SNAX streamer wrapper is trying to fix?

Reviewer: Maybe add answers to the questions in
```markdown
<details><summary>solution</summary> This is the solution! </details>
```
Rendered:
<details><summary>solution</summary> This is the solution! </details>

Author: Yes! Hold-on hehe I have to get the answers too 😅

# Try Modifying the Streamer!

Let's try a simple exercise but we will spoil the answer to you already. Suppose we want the following streamer charateristics:
- 2 temporal loop bounds
- 1 reader and 1 writer that takes in 8 parallel inputs (and outputs) each with 64 bits wide and with FIFO depth of 16.
- 8 TCDM ports for each reader and writer.

The configuration streamer template should look like:

> ```

Reviewer: ```suggestion
```hjson
```

| temporal loop bound 1          |       1         |   RW    | second temporal loop bound                           |
| temporal stride 0 for bound 0  |       2         |   RW    | Temporal stride for loop bound 0 for input           |
| temporal stride 1 for bound 0  |       3         |   RW    | Temporal stride for loop bound 0 for output          |
| temporal stride 0 for bound 1  |       4         |   RW    | Temporal stride for loop bound 1 for input           |
| temporal stride 1 for bound 1  |       5         |   RW    | Temporal stride for loop bound 1 for output          |
| spatial stride 0               |       6         |   RW    | Spatial stride for input                             |
| spatial stride 1               |       7         |   RW    | Spatial stride for output                            |        
| base addr 0                    |       8         |   RW    | Base address for input                               |
| base addr 1                    |       9         |   RW    | Base address for output                              |
| start                          |       10        |   RW    | Start register to send configurations                |

> | perf. counter                  |       11        |   RO    | Performance counter indicating number of cycles ran  | 

Reviewer: Same comments as before

Because we now have two temporal loop bounds, the temporal stride registers are also doubled. We need two temporal strides, one for input and one for output, for each temporal loop bound. Therefore there are a total of 2 temporal strides for the first loop bound and another 2 temporal strides for the second loop bound.

> Finally, the affine address that this streamer can do is:

Reviewer: ```suggestion
Finally, the affine address generation that this streamer can do is:
```
